### PR TITLE
GPOoh r conversion

### LINKED ISSUE #91: Convert GP Out of Hours extract processing

### PR DESCRIPTION
**Pull request by Jennit07**
I think this is almost ready. Definitely needs more testing/formatting but i thought i would submit the PR as i can't figure out some parts and wonder if you had any suggestions on how to write some of the code. I started to join the scripts together but then got a bit confused while testing so i've kept these separate for now and will put them into one script once finalised.

One part i am not sure of is line 103 in `extract_GPOoH_Diagnosis` in SPSS we take the first position of `.` and assign this to `diag1` to `diag6` but when i use `stringr::str_locate` i think this returns a vector which then doesn't let me do a pivot wider to create diag1 to diag6. Any thoughts on how to do this differently/improvements? This is the error returned when trying to do pivot wider:

`Error: Can't recycle ..1 (size 10) to match ..2 (size 5).
Run rlang::last_error() to see where the error occurred.`

 Still to do on this PR: 
- [ ] Create diag1 to diag6 variables with the read codes assigned accordingly by position of `.` 
- [ ] Any improvements to code where this can be simplified/run faster
- [ ] Join the 3 scripts together into one large script (kept this separate while testing)
- [ ] remove code where this writes to disk as we will store temp files as an object 
- [ ] check if there is a simpler way of doing things in R compared to SPSS code 

Closes #91

### CODE REVIEW

File: Production_scripts/Process_extracts/extract_gp_ooh_consultations.R [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+  year, ".rds"
+))
+
+# outcomes
+file.remove(paste0(
+  get_year_dir(year = year),
+  "/gp-outcomes-data-20",
+  year, ".rds"
+))
+
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_consultations.R:328]: I know in one of mine I removed the deletion of temporary files as in R you do not need to save these out, so unsure if required or not?

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  rename(
+    guid = `GUID`,
+    readcode = `Diagnosis Code`,
+    description = `Diagnosis Description`
+  )
+
+
+## Deal with Read Codes --------------------------
+
+diagnosis_clean <- diagnosis_file %>%
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:61]:
```suggestion
diagnosis_read_codes <- diagnosis_file %>%
```

rename this and leave the `diagnosis_clean` for the next section dataframe name, I got an error so it allows sections to be tested separately

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    full_match2 == 0 & readcode == "Xa1mz" ~ "S349",
+    full_match2 == 0 & readcode == "HO6.." ~ "H06..",
+    full_match2 == 0 & readcode == "zV6.." ~ "ZVz..",
+    TRUE ~ readcode
+  ))
+
+
+## Data Cleaning ---------------------------------
+
+diagnosis_clean <- diagnosis_clean %>%
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:97]:
```suggestion
diagnosis_clean <- diagnosis_read_codes %>%
```

from above

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  arrange(guid, readcode) %>%
+  # Remove duplicates (use a flag)
+  mutate(
+    duplicate = if_else(guid == lag(guid, default = first(guid)) & readcode == lag(readcode, default = first(readcode)), 1, 0)
+  ) %>%
+  filter(duplicate == 0) %>%
+  # Base R way - not working
+  # mutate(readcodelevel = gregexpr('[.]', readcode)[1])
+  mutate(
+    readcodelevel = str_locate(readcode, "[.]"),
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:108]:
```suggestion
    readcodelevel = str_locate(readcode, "[.]")[,1],
```

this creates 2 columns - numeric and description - so select the first column of this

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    readcodelevel = replace_na(readcodelevel, 0)
+  ) %>%
+  # restructure data
+  pivot_wider(
+    id_cols = guid,
+    names_from = readcodelevel,
+    values_from = readcode
+  )
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:116]: this now works with the above change

I don't know what is trying to be done to know if this works or not? 

```suggestion
  # restructure data
  pivot_wider(
   #id_cols = guid,
    names_from = readcodelevel,
    values_from = readcode
  )
```

additionally, removing this line perhaps? again I do not know the desired output but hopefully this helps get you there?

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+
+
+## Save outfile----------------------------------------
+
+outfile <- outcome_clean %>%
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:122]: what is `outcome_clean` ? is this a typo?

```suggestion
outfile <- diagnosis_clean %>%
```

File: Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+# TEMP rds file
+readr::write_rds(
+  paste0(
+    get_year_dir(year = year),
+    "/gp-outcomes-data-20",
+    year, ".rds"
+  )
+)
+
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_diagnosis.R:142]: again, potentially remove if one script as no need to save out the temp files in R?
```suggestion

```

File: Production_scripts/Process_extracts/extract_gp_ooh_outcomes.R [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+# TEMP rds file
+readr::write_rds(
+  paste0(
+    get_year_dir(year = year),
+    "/gp-outcomes-data-20",
+    year, ".rds"
+  )
+)
```
Comment (catherineholland1) [inline Production_scripts/Process_extracts/extract_gp_ooh_outcomes.R:110]:
```suggestion

```

again remove temp files if one script

**Review by catherineholland1 — CHANGES_REQUESTED** [review]
some little changes

hope I found something that might help your `pivot_wider` ? I didn't know the desired result so unsure if found the desired fix

**Comment by Jennit07** [comment]
This PR is outstanding and needs a fresh set of eyes on this. I don't think i have dealt with creating diag1 to diag6 correctly and the script can definitely be cut down in places.

**Comment by Moohan** [comment]
Just closing this while I work on it. I'll open a new PR in the future.

**Pull request closed without merging**